pipeline_model.py

```diff
@@ -92,24 +108,56 @@ def forward(self, inputs):
         shift_labels = shift_labels.view(-1)
         # Enable model parallelism
         shift_labels = shift_labels.to(shift_logits.device)
-
-        loss_unreduced = Fast_CrossEntropyLoss.apply(shift_logits, shift_labels)
-
         valid_loss = (shift_labels >= 0)
-        loss_unreduced = loss_unreduced[valid_loss]
-        optimized_loss_unreduced = loss_unreduced
-
-        # focal loss
-        if (self.focal_loss_gamma > 0):
-            p = torch.exp(-optimized_loss_unreduced)
-            optimized_loss_unreduced = (1-p)**self.focal_loss_gamma * optimized_loss_unreduced
 
+        cross_entropy_loss_unreduced = Fast_CrossEntropyLoss.apply(
+            shift_logits,
+            shift_labels,
+            self.logit_scale
+        )
+        cross_entropy_loss_unreduced = cross_entropy_loss_unreduced[valid_loss]
+
+        if self.loss_type == 'cross_entropy_loss':
+            loss_unreduced = cross_entropy_loss_unreduced
+        elif self.loss_type == 'focal_loss':
+            # See https://arxiv.org/abs/1708.02002 (Section 3)
+            p = torch.exp(-cross_entropy_loss_unreduced)
+            loss_unreduced = (1-p)**self.focal_loss_gamma * cross_entropy_loss_unreduced
+        elif self.loss_type == 'focal_loss_star':
```

Comment:
> In a comment you said that logit_scale needs to be moved here for the focal_loss_star implementation. Why? If the logits are already scaled by logit_scale (from the model code), can't you just scale it again by focal_loss_gamma when passing it into the cross entropy loss function? If the worry is that you would be making another copy of the logits tensor (which may be large), you could still have the cross entropy loss support passing in a logit scale parameter for additional scaling.
> 
> If possible, I like having logit_scale baked into the model itself, since conceptually it's part of the model. For example, if someone were to adapt the pipeline parallel wrappers to do inference using a PipelineModule, it wouldn't work if you were relying on the loss function to handle the logit scale. The actual logits output by the model need to be scaled according to the logit_scale in the model config.

Yeah, that's a good point.

I did originally try scaling the logits that went into the cross-entropy loss and it did use a lot of extra VRAM.

I will revert the changes to the Cohere model and just use the `logit_scale` parameter for Focal Loss* tomorrow.

Reply:
I've reverted the changes for Cohere, but just testing it now on `command-r:35b` so won't close this comment until I see it's OK.

Reply:
Yeah, it looks fine - losses and Entropy back to same.